File: 281 Zigzag Iterator/zigzag_iterator.cpp
class ZigzagIterator {	// Author: Xinyu Chen
public:
	ZigzagIterator(std::vector<int>& v1, std::vector<int>& v2) {
		auto i = v1.begin(), j = v2.begin();
			while (i != v1.end() && j != v2.end()) {
				data.push_back(*i); ++i;
				data.push_back(*j); ++j;
			}
		if (i == v1.end()) {
			for (; j != v2.end(); ++j) data.push_back(*j);
		}
		else {
			for (; i != v1.end(); ++i) data.push_back(*i);
		}
		iter = data.begin();
	}

	int next() {
		int ret = *iter;
		++iter;
		return ret;
	}

	bool hasNext() {
		return iter != data.end();
	}
private:
	std::vector<int> data;
	std::vector<int>::iterator iter;
};

/**
 * Your ZigzagIterator object will be instantiated and called as such:
 * ZigzagIterator i(v1, v2);
 * while (i.hasNext()) cout << i.next();
 */
 

File: 281 Zigzag Iterator/zigzag_iterator.java
public class ZigzagIterator {
    int idx1, idx2;
    List<Integer> v1, v2;
    boolean isFirst;
    public ZigzagIterator(List<Integer> v1, List<Integer> v2) {
        idx1 = 0; 
        idx2 = 0;
        this.v1 = v1;
        this.v2 = v2;
        isFirst = v1.size() == 0 ? false : true;
    }

    public int next() {  
        if (isFirst) {
            if (idx2 != v2.size()) isFirst = !isFirst;
            return v1.get(idx1++);
        } else {
            if (idx1 != v1.size()) isFirst = !isFirst;
            return v2.get(idx2++);
        }
    }

    public boolean hasNext() {
        return idx1 != v1.size() || idx2 != v2.size();
    }
} // Author: Cassie

/**
 * Your ZigzagIterator object will be instantiated and called as such:
 * ZigzagIterator i = new ZigzagIterator(v1, v2);
 * while (i.hasNext()) v[f()] = i.next();
 */
